limacharlie/firehose.go

```diff
@@ -0,0 +1,308 @@
+	}
+	if opts.Tag != nil {
+		output.Tag = *opts.Tag
+	}
+	if opts.Category != nil {
+		output.Category = *opts.Category
+	}
+	if opts.IsDeleteOnFailure != nil {
+		output.DeleteOnFailure = *opts.IsDeleteOnFailure
+	}
+	return output
+}
+
+type FirehoseOptions struct {
+	ListenOnPort         uint16
+	ListenOnIP           net.IP
+	ConnectToPort        uint16
+	ConnectToIP          net.IP
+	SSLCertPath          string
+	SSLCertKeyPath       string
+	MaxMessageCount      int
+	MaxErrorMessageCount int
+	InvestigationID      string
+	EventTag             string
+	DetectionCategory    string
+	SensorID             string
+	DeleteOnFailure      bool
+	ParseMessage         bool
+}
+
+type FirehoseMessage struct {
+	Message string
+}
+
+type Firehose struct {
+	Organization     Organization
+	opts             FirehoseOptions
+	outputOpts       *FirehoseOutputOptions // optional
+	Messages         chan FirehoseMessage
+	ErrorMessages    chan FirehoseMessage
+	messageDropCount int
+	listener         net.Listener
+	shutdownMessage  chan bool
+}
+
+type firehoseHandler struct {
+	Options FirehoseOptions
+}
+
+func createSelfSignedCertificate() (*tls.Certificate, error) {
+	certPrivateKey, err := rsa.GenerateKey(rand.Reader, 4096)
+	if err != nil {
+		return nil, err
+	}
+	certTemplate := x509.Certificate{
+		SerialNumber: big.NewInt(1),
+		Subject: pkix.Name{
+			CommonName:    "limacharlie_firehose",
+			Organization:  []string{"refractionPOINT"},
+			Locality:      []string{"Mountain View"},
+			Province:      []string{"CA"},
+			Country:       []string{"US"},
+			StreetAddress: []string{""},
+			PostalCode:    []string{""},
+		},
+		IPAddresses: []net.IP{net.IPv4(127, 0, 0, 1), net.IPv6loopback},
+		NotBefore:   time.Now(),
+		NotAfter:    time.Now().AddDate(10, 0, 0),
+		ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageAny},
+		KeyUsage:    x509.KeyUsageDigitalSignature,
+		IsCA:        true,
+	}
+	certBytes, err := x509.CreateCertificate(rand.Reader, &certTemplate, &certTemplate, &certPrivateKey.PublicKey, certPrivateKey)
+	if err != nil {
+		return nil, err
+	}
+	certPEM := new(bytes.Buffer)
+	pem.Encode(certPEM, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: certBytes,
+	})
+	certPrivKeyPEM := new(bytes.Buffer)
+	pem.Encode(certPrivKeyPEM, &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(certPrivateKey),
+	})
+	cert, err := tls.X509KeyPair(certPEM.Bytes(), certPrivKeyPEM.Bytes())
+	if err != nil {
+		return nil, err
+	}
+	return &cert, nil
+}
+
+func startListener(listenOnIP net.IP, listenOnPort uint16, sslCertPath string, sslCertKeyPath string) (*net.Listener, error) {
+	createTempCert := len(sslCertPath) == 0
+	createTempKey := len(sslCertKeyPath) == 0
+	if createTempCert && !createTempKey {
+		return nil, fmt.Errorf("certificate key path missing")
+	}
+	if !createTempCert && createTempKey {
+		return nil, fmt.Errorf("certificate path missing")
+	}
+
+	var certificate *tls.Certificate = nil
```

Comment:
The `= nil` is redundant here as the default value of a var that's a pointer is nil.